docs/operators/deduper.md

```diff
@@ -0,0 +1,523 @@
+be two tuples t1 and t2 such that:
+
+Tuple 1: d1, e1
+
+Tuple 2: d2, e2
+
+d1 = d2 and e1 != e2
+
+This assumption was made with respect to certain use cases. These
+use cases follow this assumption in that the records which are
+duplicates are exactly identical. An example use case is when log
+messages are replayed erroneously, and we want to identify the duplicate
+log messages. In such cases, we need not worry about two different log
+messages having the same identifier but different timestamps. Since its
+a replay of the same data, the duplicate records are assumed to be
+exactly identical.
+
+Flow of a Tuple through Dedup Operator
+---------------------------------------------------
+
+Tuples flow through the Dedup operator one by one. Deduper may
+choose to process a tuple immediately, or store it in some data
+structure for later processing. We break down the processing of the
+Deduper by various stages as follows.
+
+### Deduper View
+
+A tuple always arrives at the input
+port of the Dedup operator. Once arrived, the Deduper does
+the following tasks.
+
+#### Check if tuple is Expired
+
+This is only applicable in case of Dedup with expiry. The
+following condition can be used to check if the tuple is expired.
+
+if ( Latest Point - Expiry Key \< Expiry Point ) then
+Expired
+
+If the tuple is expired, then put it out to the expired port.
+
+#### Check if tuple is a Duplicate or Unique
+
+Once a tuple passes the check of expiry, we proceed to check if
+the tuple already has a duplicate tuple which is not expired. Note that
+if the tuple in question is not expired, the duplicate will also not
+have expired due to the assumption listed [here](#Assumption1). The Deduper queries the
+Managed state to fetch the bucket for the tuple. This request is
+processed by the Managed state in a separate asynchronous thread. Once
+this request is submitted, the Deduper moves on to process other
+tuples.
+
+Once the request to Managed state is completed and the bucket is
+fetched from persistent storage,
+
+Additionally the Deduper also inserts the tuple being processed
+into a waiting events queue for
+later processing. After this, the Deduper cannot proceed until the
+bucket is loaded by the Bucket Manager and hence proceeds to process
+another tuple. Next section details the process after the bucket is
+loaded by the Bucket Manager.
+
+Ports, Attributes and Properties
+==========================================================
+
+Ports
+------------------
+
+The deduper has a single input port and multiple output
+ports.
+
+
+
+-   ***input*** - This is the input port through
+    which the tuples arrive at the Deduper.
+-   ***unique*** - This is the output port on
+    which unique tuples are sent out by the Deduper.
+-   ***duplicate*** - This is the output port on
+    which duplicate tuples are sent out by the Deduper.
+-   ***expired*** - This is the output port on
+    which expired tuples are sent out by the Deduper.
```

Comment:
Worth mentioning if which are optional and which required for users who don't care to handle expired tuples for example.